# Dependencies

In [1]:
import keras 
from keras import layers
from keras.layers import Input, Dense, Conv2D, MaxPooling2D, UpSampling2D, Dropout, Flatten
from keras.models import Model, Sequential
from keras.datasets import mnist
from keras import backend as K
from keras.preprocessing.image import img_to_array
from keras.applications.vgg19 import VGG19, preprocess_input
from keras.utils import np_utils

import sklearn
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV, train_test_split, cross_validate
from sklearn.metrics import confusion_matrix, accuracy_score, auc, roc_auc_score, roc_curve
from sklearn.externals import joblib
from sklearn.preprocessing import LabelEncoder

from random import shuffle

import numpy as np
import pandas as pd
import matplotlib
%matplotlib inline
import matplotlib.pyplot as plt

Using TensorFlow backend.
/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarn

In [2]:
#Import the MNIST dataset
(X_train, _), (X_test, _) = mnist.load_data()

In [3]:
X_train = X_train.astype('float32')/255.0
X_test = X_test.astype('float32')/255.0

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test2 = train_test_split(X_train, test_size=0.8, random_state=42)

In [5]:
print("Training data shape: ", X_train.shape)

print("Testing data shape: ", X_test.shape)

Training data shape:  (12000, 28, 28)
Testing data shape:  (10000, 28, 28)


# Building a model 

In [6]:
#Shape the data

X_train_MLP = X_train.reshape((X_train.shape[0], -1))
X_test_MLP = X_test.reshape((X_test.shape[0], -1))
print("Training data_MLP shape: ", X_train_MLP.shape)
print("Testing data_MLP shape: ", X_test_MLP.shape)

Training data_MLP shape:  (12000, 784)
Testing data_MLP shape:  (10000, 784)


In [7]:
# For the purpose of consitancy, the complile and fit are same across the notebook 

# One hidden layer

In [8]:
INPUT_SIZE = 784
HIDDEN_ONE = 64
ENCODING_SIZE = 32

input_MLP = Input(shape=(INPUT_SIZE,))

MLP_encode = Dense(HIDDEN_ONE, activation='relu')(input_MLP)

MLP_encode = Dense(ENCODING_SIZE, activation='relu')(MLP_encode)

MLP_decode = Dense(HIDDEN_ONE, activation='relu')(MLP_encode)
MLP_decode = Dense(INPUT_SIZE, activation='relu')(MLP_decode)

autoencoder_MLP_1 = Model(input_MLP, MLP_decode)

In [9]:
autoencoder_MLP_1.compile(optimizer='adam',loss='mean_squared_error')

autoencoder_MLP_1.fit(X_train_MLP,
                   X_train_MLP,
                   epochs=10,
                   batch_size = 64,
                   shuffle=True,
                   validation_split=0.2)

decoded_MLP_1 = autoencoder_MLP_1.predict(X_test_MLP)


Train on 9600 samples, validate on 2400 samples
Epoch 1/10
9600/9600 [==============================] - 0s 42us/step - loss: 0.0539 - val_loss: 0.0339
Epoch 2/10
9600/9600 [==============================] - 0s 25us/step - loss: 0.0299 - val_loss: 0.0270
Epoch 3/10
9600/9600 [==============================] - 0s 24us/step - loss: 0.0254 - val_loss: 0.0240
Epoch 4/10
9600/9600 [==============================] - 0s 25us/step - loss: 0.0232 - val_loss: 0.0226
Epoch 5/10
9600/9600 [==============================] - 0s 24us/step - loss: 0.0220 - val_loss: 0.0218
Epoch 6/10
9600/9600 [==============================] - 0s 26us/step - loss: 0.0212 - val_loss: 0.0210
Epoch 7/10
9600/9600 [==============================] - 0s 24us/step - loss: 0.0206 - val_loss: 0.0205
Epoch 8/10
9600/9600 [==============================] - 0s 25us/step - loss: 0.0202 - val_loss: 0.0203
Epoch 9/10
9600/9600 [==============================] - 0s 24us/step - loss: 0.0198 - val_loss: 0.0200
Epoch 10/10
9600/9600 [=

# Two Hidden Layer

In [10]:
INPUT_SIZE = 784
HIDDEN_ONE = 128
HIDDEN_TWO = 64
ENCODING_SIZE = 32

input_MLP = Input(shape=(INPUT_SIZE,))

MLP_encode = Dense(HIDDEN_ONE, activation='relu')(input_MLP)
MLP_encode = Dense(HIDDEN_TWO, activation='relu')(MLP_encode)

MLP_encode = Dense(ENCODING_SIZE, activation='relu')(MLP_encode)

MLP_decode = Dense(HIDDEN_TWO, activation='relu')(MLP_encode)
MLP_decode = Dense(HIDDEN_ONE, activation='relu')(MLP_decode)
MLP_decode = Dense(INPUT_SIZE, activation='relu')(MLP_decode)

autoencoder_MLP_2 = Model(input_MLP, MLP_decode)

In [11]:
autoencoder_MLP_2.compile(optimizer='adam',loss='mean_squared_error')

autoencoder_MLP_2.fit(X_train_MLP,
                   X_train_MLP,
                   epochs=10,
                   batch_size=64,
                   shuffle=True,
                   validation_split=0.2)

decoded_MLP_2 = autoencoder_MLP_2.predict(X_test_MLP)

Train on 9600 samples, validate on 2400 samples
Epoch 1/10
9600/9600 [==============================] - 1s 64us/step - loss: 0.0524 - val_loss: 0.0339
Epoch 2/10
9600/9600 [==============================] - 0s 41us/step - loss: 0.0300 - val_loss: 0.0274
Epoch 3/10
9600/9600 [==============================] - 0s 38us/step - loss: 0.0258 - val_loss: 0.0245
Epoch 4/10
9600/9600 [==============================] - 0s 39us/step - loss: 0.0236 - val_loss: 0.0231
Epoch 5/10
9600/9600 [==============================] - 0s 37us/step - loss: 0.0224 - val_loss: 0.0221
Epoch 6/10
9600/9600 [==============================] - 0s 39us/step - loss: 0.0214 - val_loss: 0.0213
Epoch 7/10
9600/9600 [==============================] - 0s 38us/step - loss: 0.0206 - val_loss: 0.0207
Epoch 8/10
9600/9600 [==============================] - 0s 37us/step - loss: 0.0201 - val_loss: 0.0201
Epoch 9/10
9600/9600 [==============================] - 0s 38us/step - loss: 0.0195 - val_loss: 0.0197
Epoch 10/10
9600/9600 [==

# Three Hidden Layers

In [12]:
INPUT_SIZE = 784
HIDDEN_ONE = 256
HIDDEN_TWO = 128
HIDDEN_THREE = 64
ENCODING_SIZE = 32

input_MLP = Input(shape=(INPUT_SIZE,))
MLP_encode = Dense(HIDDEN_ONE, activation='relu')(input_MLP)
MLP_encode = Dense(HIDDEN_TWO, activation='relu')(MLP_encode)
MLP_encode = Dense(HIDDEN_THREE, activation='relu')(MLP_encode)

MLP_encode = Dense(ENCODING_SIZE, activation='relu')(MLP_encode)

MLP_decode = Dense(HIDDEN_THREE, activation='relu')(MLP_encode)
MLP_decode = Dense(HIDDEN_TWO, activation='relu')(MLP_decode)
MLP_decode = Dense(HIDDEN_ONE, activation='relu')(MLP_decode)
MLP_decode = Dense(INPUT_SIZE, activation='relu')(MLP_decode)
autoencoder_MLP_3 = Model(input_MLP, MLP_decode)

In [13]:
autoencoder_MLP_3.compile(optimizer='adam',loss='mean_squared_error')

autoencoder_MLP_3.fit(X_train_MLP,
                   X_train_MLP,
                   epochs=10,
                   batch_size=64,
                   shuffle=True,
                   validation_split=0.2)

decoded_MLP_3 = autoencoder_MLP_3.predict(X_test_MLP)

Train on 9600 samples, validate on 2400 samples
Epoch 1/10
9600/9600 [==============================] - 1s 96us/step - loss: 0.0578 - val_loss: 0.0372
Epoch 2/10
9600/9600 [==============================] - 1s 62us/step - loss: 0.0323 - val_loss: 0.0293
Epoch 3/10
9600/9600 [==============================] - 1s 64us/step - loss: 0.0272 - val_loss: 0.0253
Epoch 4/10
9600/9600 [==============================] - 1s 63us/step - loss: 0.0241 - val_loss: 0.0235
Epoch 5/10
9600/9600 [==============================] - 1s 66us/step - loss: 0.0221 - val_loss: 0.0217
Epoch 6/10
9600/9600 [==============================] - 1s 69us/step - loss: 0.0207 - val_loss: 0.0206
Epoch 7/10
9600/9600 [==============================] - 1s 67us/step - loss: 0.0198 - val_loss: 0.0198
Epoch 8/10
9600/9600 [==============================] - 1s 68us/step - loss: 0.0190 - val_loss: 0.0191
Epoch 9/10
9600/9600 [==============================] - 1s 69us/step - loss: 0.0184 - val_loss: 0.0186
Epoch 10/10
9600/9600 [==

# Two Hidden Layer With Chanegs
  
# Increasing the size of the hidden layers

In [14]:
INPUT_SIZE = 784
HIDDEN_ONE = 512 #increasing the size of the layer
HIDDEN_TWO = 256 #increasing the size of the layer
ENCODING_SIZE = 32

input_MLP = Input(shape=(INPUT_SIZE,))

MLP_encode = Dense(HIDDEN_ONE, activation='relu')(input_MLP)
MLP_encode = Dense(HIDDEN_TWO, activation='relu')(MLP_encode)

MLP_encode = Dense(ENCODING_SIZE, activation='relu')(MLP_encode)

MLP_decode = Dense(HIDDEN_TWO, activation='relu')(MLP_encode)
MLP_decode = Dense(HIDDEN_ONE, activation='relu')(MLP_decode)
MLP_decode = Dense(INPUT_SIZE, activation='relu')(MLP_decode)

autoencoder_MLP_2_Change = Model(input_MLP, MLP_decode)

In [15]:
autoencoder_MLP_2_Change.compile(optimizer='adam',loss='mean_squared_error')

autoencoder_MLP_2_Change.fit(X_train_MLP,
                   X_train_MLP,
                   epochs=10,
                   batch_size=64,
                   shuffle=True,
                   validation_split=0.2)

decoded_MLP_2_Changes = autoencoder_MLP_2_Change.predict(X_test_MLP)

Train on 9600 samples, validate on 2400 samples
Epoch 1/10
9600/9600 [==============================] - 2s 162us/step - loss: 0.0412 - val_loss: 0.0255
Epoch 2/10
9600/9600 [==============================] - 1s 130us/step - loss: 0.0229 - val_loss: 0.0212
Epoch 3/10
9600/9600 [==============================] - 1s 137us/step - loss: 0.0195 - val_loss: 0.0188
Epoch 4/10
9600/9600 [==============================] - 1s 142us/step - loss: 0.0178 - val_loss: 0.0178
Epoch 5/10
9600/9600 [==============================] - 1s 143us/step - loss: 0.0167 - val_loss: 0.0168
Epoch 6/10
9600/9600 [==============================] - 1s 142us/step - loss: 0.0159 - val_loss: 0.0163
Epoch 7/10
9600/9600 [==============================] - 1s 136us/step - loss: 0.0153 - val_loss: 0.0160
Epoch 8/10
9600/9600 [==============================] - 1s 144us/step - loss: 0.0148 - val_loss: 0.0156
Epoch 9/10
9600/9600 [==============================] - 1s 148us/step - loss: 0.0144 - val_loss: 0.0153
Epoch 10/10
9600

# Outputs 

    One Hidden Layer      loss: 0.0194 - val_loss: 0.0193
  
    Two Hidden Layer       loss: 0.0190 - val_loss: 0.0192
      
    Three Hidden Layer     oss: 0.0179 - val_loss: 0.0185
  
    Two Hidden Layer With Chanegs   loss: 0.0141 - val_loss: 0.0149

In [28]:
# We can see that, either increasing the number of hidden layers or the number of nodes in a layer, accuracy improves.

# Quesiton 2
# Build any two CNN architectures 

In [16]:
batch_size = 128
num_classes = 10 # 0-9 
epochs = 3

# input image dimensions
img_rows, img_cols = 28, 28

In [17]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, ignore, y_train, ignore2 = train_test_split(x_train, y_train, test_size=0.8, random_state=42)


if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

x_train shape: (12000, 28, 28, 1)
12000 train samples
10000 test samples


# Model 1

In [18]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

In [19]:
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

In [20]:
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

Train on 12000 samples, validate on 10000 samples
Epoch 1/3
12000/12000 [==============================] - 13s 1ms/step - loss: 0.6547 - accuracy: 0.7924 - val_loss: 0.1655 - val_accuracy: 0.9533
Epoch 2/3
12000/12000 [==============================] - 13s 1ms/step - loss: 0.2000 - accuracy: 0.9385 - val_loss: 0.1128 - val_accuracy: 0.9636
Epoch 3/3
12000/12000 [==============================] - 13s 1ms/step - loss: 0.1357 - accuracy: 0.9588 - val_loss: 0.0741 - val_accuracy: 0.9756


In [21]:
from keras import models
#if loaded_model:
#    model = loaded_model
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1]) 

Test loss: 0.07407280763722957
Test accuracy: 0.975600004196167


# MODEL 2

In [22]:
#Changes made for Model 2:
# Reversed the size of Layers 
# Removed The dropout layers
# Changed Kernel_size 
model2 = Sequential()
model2.add(Conv2D(128, kernel_size=(4, 4),
                 activation='relu',
                 input_shape=input_shape))
model2.add(Conv2D(64, (3, 3), activation='relu'))
model2.add(MaxPooling2D(pool_size=(2, 2)))

model2.add(Flatten())
model2.add(Dense(32, activation='relu'))

model2.add(Dense(num_classes, activation='softmax'))

In [23]:
model2.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

In [24]:
model2.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

Train on 12000 samples, validate on 10000 samples
Epoch 1/3
12000/12000 [==============================] - 27s 2ms/step - loss: 0.6803 - accuracy: 0.7881 - val_loss: 0.1858 - val_accuracy: 0.9422
Epoch 2/3
12000/12000 [==============================] - 26s 2ms/step - loss: 0.1615 - accuracy: 0.9534 - val_loss: 0.1127 - val_accuracy: 0.9623
Epoch 3/3
12000/12000 [==============================] - 26s 2ms/step - loss: 0.0899 - accuracy: 0.9726 - val_loss: 0.0978 - val_accuracy: 0.9692


In [25]:
score = model2.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1]) 

Test loss: 0.09783693803958594
Test accuracy: 0.9692000150680542


Model 1:  Test accuracy: 0.9746999740600586  
Model 2: Test accuracy: 0.9732000231742859


In [26]:
# We can see that removing the dropout layers does affect the Accuracy of the model.